src/Microsoft.ComponentDetection.Detectors/pip/IPyPiClient.cs

```diff
@@ -138,61 +143,66 @@ public async Task<IList<PipDependencySpecification>> FetchPackageDependenciesAsy
-                }
-
-                var statusCode = (int)message.StatusCode;
-
-                // only retry if server doesn't classify the call as a client error!
-                var isRetryable = statusCode < 400 || statusCode > 499;
-                return !message.IsSuccessStatusCode && isRetryable;
-            })
-            .WaitAndRetryAsync(1, i => RETRYDELAY, (result, timeSpan, retryCount, context) =>
-            {
-                using var r = new PypiRetryTelemetryRecord { Name = spec.Name, DependencySpecifiers = spec.DependencySpecifiers?.ToArray(), StatusCode = result.Result.StatusCode };
-
-                this.logger.LogWarning(
-                    "Received {StatusCode} {ReasonPhrase} from {RequestUri}. Waiting {TimeSpan} before retry attempt {RetryCount}",
-                    result.Result.StatusCode,
-                    result.Result.ReasonPhrase,
-                    requestUri,
-                    timeSpan,
-                    retryCount);
-
-                Interlocked.Increment(ref this.retries);
-            })
-            .ExecuteAsync(() =>
-            {
-                if (Interlocked.Read(ref this.retries) >= MAXRETRIES)
+                    // stop retrying if MAXRETRIES was hit!
+                    if (message == null)
+                    {
+                        return false;
+                    }
+
+                    var statusCode = (int)message.StatusCode;
+
+                    // only retry if server doesn't classify the call as a client error!
+                    var isRetryable = statusCode < 400 || statusCode > 499;
+                    return !message.IsSuccessStatusCode && isRetryable;
+                })
+                .WaitAndRetryAsync(1, i => RETRYDELAY, (result, timeSpan, retryCount, context) =>
                 {
-                    return Task.FromResult<HttpResponseMessage>(null);
-                }
-
-                return this.GetAndCachePyPiResponseAsync(requestUri);
-            });
+                    using var r = new PypiRetryTelemetryRecord { Name = spec.Name, DependencySpecifiers = spec.DependencySpecifiers?.ToArray(), StatusCode = result.Result.StatusCode };
+
+                    this.logger.LogWarning(
+                        "Received {StatusCode} {ReasonPhrase} from {RequestUri}. Waiting {TimeSpan} before retry attempt {RetryCount}",
+                        result.Result.StatusCode,
+                        result.Result.ReasonPhrase,
+                        requestUri,
+                        timeSpan,
+                        retryCount);
+
+                    Interlocked.Increment(ref this.retries);
+                })
+                .ExecuteAsync(() =>
+                {
+                    if (Interlocked.Read(ref this.retries) >= MAXRETRIES)
+                    {
+                        return Task.FromResult<HttpResponseMessage>(null);
+                    }
+
+                    return this.GetAndCachePyPiResponseAsync(requestUri);
```

Comment:
is this bypassing the cache?

Reply:
yes, it should be adding to cached instead. fixed it.

Reply:
actually I think I misunderstood your question. do you mean is the shared cache bypassing this internal cache? yes it is, i'm checking the shared cache first and if it returns nothing, then it enters this retry block and tries to get from the internal cache and if that returns nothing, it sends a request to the pypi api.